designs/2023-06-04-exec-platform-scoped-spawn-strategies.md

```diff
@@ -0,0 +1,154 @@
+  `--strategy_regexp=//foo.*\.cc,-//foo/bar=local`
+
+Collectively the flags and available spawn strategies offer a great deal of control, but not enough to cover all multi-platform build scenarios.
+
+For example;
+
+```starlark
+# //BUILD.bazel
+
+# Unsupported on remote
+platform(
+    name = "darwin_arm64",
+    ...
+)
+
+platform(
+    name = "linux_amd64",
+    ...
+)
+
+platform(
+    name = "win_amd64",
+    ...
+)
+
+foo_binary(
+    name = "foo-bin"
+)
+
+```
+
+```ini
+# //.bazelrc
+build --spawn_strategy=remote
+build --host_platform=//:darwin_amd64
+build --extra_exec_platforms=...
+# extra bits for remote connection
+```
+
+```sh
+# Shell at //
+bazel build //:foo-bin --platforms=//:darwin_arm64,//:linux_amd64,//:win_amd64
+```
+
+In this scenario _all_ actions will be executed on the remote or pulled from its cache. If `//:foo-bin` has previously been built with the exact same inputs (and configuration???) the build will pass thanks to the cache hit, otherwise the build will fail (e.g. attempted to run darwin executable on Linux or found no suitable executor).
+
+There are ways to work around this problem, but all have costs.
+- ...separate builds for platforms not buildable with remote...
+- ...somehow customise mnemonics or description...
+
+For this trivial scenario the challenges can be overcome with relative ease, however this does not scale well. ...dependencies...
+
+
+# Proposal
+
+<!-- The actual proposal. This should be detailed enough for users to understand the
+change, its motivations, and the rationale. When appropriate, include rejected
+alternatives (and why they were rejected).
+
+The current file is a template. Copy it, update it to your needs. Feel free to add or
+remove sections. -->
+Support scoping of existing spawn strategy flags to a specific execution platform.
+
+e.g.
+
+```ini
+# //.bazelrc
+build --spawn_strategy=//:darwin_arm64=worker,sandboxed,local
+build --strategy=FOO=//:darwin_arm64=local
+build --strategy_regexp=//foo.*\.cc=//:darwin_arm64=local
+```
+
+The motivation behind this approach is;
+1. Implementation simplicity. A prototype should not require any significant changes to Bazel internals.
+2. Existing strategy flags are easy for scripts to interact with. e.g. an opt-in remote support script can be run before a build ensure a tunnel is opened and write out of a `.bazelrc` file which is read with `try-import`.
+
+...
+
+The likihood of misconfiguration can be reduced with;
+- `--[no]require_platform_scoped_strategies` to make inclusion of platforms in spawn strategy flags mandatory.
+- `--[no]exhaustive_platform_strategies` to make specifiction of spawn strategies for every registered execution platform required.
+
+
+## Priority
+
+Currently;
+- Description (`--stategy_regexp`), unless mnemonic is `TestRunner`.
+- Mnemonic (`--strategy`).
+- Defaults (`--spawn_strategy` or Bazel generated defaults).
+
+This will become;
+- Platform + description.
```

Comment:
In the current prototype the priority remains unchanged (each strategy is considered in-order). I'm considering changing this proposal to match as preserving existing behaviour seems more ergonomic.